src/wabt.cpp

```diff
@@ -116,112 +404,724 @@ ExecutionResult WabtEngine::execute(
+      interp::TypedValues&
+    ) {
+      interface.eeiSelfDestruct(args[0].value.i32);
+      return interp::Result::Ok;
+    }
+  );
+
+  hostModule->AppendFuncExport(
+    "getBlockTimestamp",
+    {{}, {Type::I64}},
+    [&interface](
+      const interp::HostFunc*,
+      const interp::FuncSignature*,
+      const interp::TypedValues&,
+      interp::TypedValues& results
+    ) {
+      results[0].set_i64(static_cast<uint64_t>(interface.eeiGetBlockTimestamp()));
+      return interp::Result::Ok;
+    }
+  );
+
+  // some extra methods for debugging
+
+#if HERA_DEBUGGING
+  hostModule->AppendFuncExport(
+    "printMemHex",
+    {{Type::I32, Type::I32}, {}},
+    [&interface](
+      const interp::HostFunc*,
+      const interp::FuncSignature*,
+      const interp::TypedValues& args,
+      interp::TypedValues&
+    ) {
+      interface.debugPrintMem(1, args[0].value.i32, args[1].value.i32);
+      return interp::Result::Ok;
+    }
+  );
+
+  hostModule->AppendFuncExport(
+    "printStorageHex",
+    {{Type::I32, Type::I32}, {}},
+    [&interface](
+      const interp::HostFunc*,
+      const interp::FuncSignature*,
+      const interp::TypedValues& args,
+      interp::TypedValues&
+    ) {
+      interface.debugPrintStorage(1, args[0].value.i32);
+      return interp::Result::Ok;
+    }
+  );
+#endif
+
+  // Parse module
+  ReadBinaryOptions options(
+    Features{},
+    nullptr, // debugging stream for loading
+    false, // ReadDebugNames
+    true, // StopOnFirstError
+    true // FailOnCustomSectionError
+  );
+
+  Errors errors;
+  interp::DefinedModule* module = nullptr;
+  Result loadResult = ReadBinaryInterp(
+    &env,
+    code.data(),
+    code.size(),
+    options,
+    &errors,
+    &module
+  );
+
+  ensureCondition(Succeeded(loadResult) && module, ContractValidationFailure, "Module failed to load.");
+  ensureCondition(env.GetMemoryCount() == 1, ContractValidationFailure, "Multiple memory sections exported.");
+  ensureCondition(module->start_func_index == kInvalidIndex, ContractValidationFailure, "Contract contains start function.");
+
+  // Prepare to execute
+  interp::Export* mainFunction = module->GetExport("main");
+  ensureCondition(mainFunction, ContractValidationFailure, "\"main\" not found");
+  ensureCondition(mainFunction->kind == ExternalKind::Func, ContractValidationFailure,  "\"main\" is not a function");
+  interp::Executor executor(
+    &env,
+    nullptr, // null for no tracing
+    interp::Thread::Options{} // empty for no threads
+  );
+
+  ensureCondition(module->GetExport("memory")!=nullptr, ContractValidationFailure, "\"memory\" not found");
+
+  // FIXME: really bad design
+  interface.setWasmMemory(env.GetMemory(0));
+  // Execute main
+  try {
+    interp::ExecResult wabtResult = executor.RunExport(mainFunction, interp::TypedValues{}); // second arg is empty since no args
+    ensureCondition((int)wabtResult.result == 0, VMTrap, "The VM invokaction had a trap.");
```

Comment:
Just checked, `.result` is an enum class. I don't think typecasting is the best option here, but checking against `Result::Ok` seems to be better.